fastly/observability_custom_dashboard.go

```diff
@@ -0,0 +1,275 @@
+
+import (
+	"encoding/json"
+	"net/http"
+	"strconv"
+	"time"
+)
+
+const path = "/observability/dashboards"
+
+const (
+	ErrInvalidDataSourceType    = ""
+	ErrInvalidMetric            = ""
+	ErrInvalidVisualizationType = ""
+	ErrInvalidPlotType          = ""
+	ErrInvalidCalculationMethod = ""
+)
+
+// ObservabilityCustomDashboard is a named container for a custom dashboard configuration
+type ObservabilityCustomDashboard struct {
+	// The date and time the dashboard was created
+	CreatedAt time.Time
+	// The ID of the user who created the dashboard
+	CreatedBy string
+	// A short description of the dashboard
+	Description string
+	// The unique identifier of the dashboard
+	ID string
+	// A list of DashboardItems
+	Items []DashboardItem
+	// A human-readable name
+	Name string
+	// The date and time the dashboard was last updated
+	UpdatedAt time.Time
+	// The ID of the user who last modified the dashboard
+	UpdatedBy string
+}
+
+// DashboardItem describes an item (or "widget") of a dashboard
+type DashboardItem struct {
+	// DataSource describes the source of the metrics to be displayed (required)
+	DataSource DashboardDataSource `json:"data_source"`
+	// ID is a unique identifier for the DashboardItem (read-only)
+	ID string `json:"id,omitempty"`
+	// Span is the number of columns (1-12) for the DashboardItem to span (default: 4)
+	Span uint8 `json:"span"`
+	// Subtitle is a human-readable subtitle to display, often a description of the visualization (optional)
+	Subtitle string `json:"subtitle"`
+	// Title is a human-readable title to display (optional)
+	Title string `json:"title"`
+	// Visualization describes the way the DashboardItem should display data (required)
+	Visualization DashboardVisualization `json:"visualization"`
+}
+
+type DashboardSourceType string
+
+const (
+	SourceTypeStatsEdge   = "stats.edge"
+	SourceTypeStatsDomain = "stats.domain"
+	SourceTypeStatsOrigin = "stats.origin"
+)
+
+// DashboardDataSource describes the data to display in a DashboardItem
+type DashboardDataSource struct {
+	// Config describes configuration options for the selected data source (required)
+	Config DashboardSourceConfig `json:"config"`
+	// Type is the source of the data to display (required)
+	Type DashboardSourceType `json:"type"`
+}
+
+type DashboardSourceConfig struct {
+	// Metrics is the list metrics to visualize (required)
+	// Valid options are defined by the selected SourceType. See https://www.fastly.com/documentation/reference/api/observability/custom-dashboards/#data-source
+	Metrics []string `json:"metrics"`
+}
+
+type VisualizationType string
+
+const VisualizationTypeChart VisualizationType = "chart"
+
+type DashboardVisualization struct {
+	// Config describes configuration options for the given visualization (required)
+	Config VisualizationConfig `json:"config"`
+	// Type is type of visualization to display. Currently only "chart" is supported (required)
+	Type VisualizationType `json:"type"`
+}
+
+type PlotType string
```

Comment:
@kpfleming I considered making these _types_ private, thereby forcing the user to use our exported constants, but I decided against it so that users aren't forced to upgrade the client in order to take advantage of future new functionality in the API.